File: packages/waddle_shared/lib/persistence/overlay_type_label.dart
import 'tables.dart';

/// Human-facing labels for overlay rows / [OverlayTypes] rows.
const Map<String, String> kOverlayTypeTitles = {
  kOverlayTypeShapeRain: 'Shape rain',
  kOverlayTypeBirthdayConfetti: 'Birthday confetti',
  kOverlayTypeBouncingMessage: 'Bouncing message',
  kOverlayTypeFallingImages: 'Falling images',
  kOverlayTypeMatrixRain: 'Matrix rain',
  kOverlayTypeEdgeGlow: 'Edge glow',
  kOverlayTypeFloatingBalloons: 'Floating balloons',
};

String _capitalizeToken(String word) {
  if (word.isEmpty) return word;
  return '${word[0].toUpperCase()}${word.substring(1).toLowerCase()}';
}

String _titleFromSegments(String overlayType) {
  final parts = overlayType.split('_').where((s) => s.isNotEmpty).toList();
  if (parts.isEmpty) return overlayType;
  return parts.map(_capitalizeToken).join(' ');
}

/// Normalized label for cards, dialogs, and [OverlayTypes.label] seeding.
String overlayTypeLabel(String overlayType) {
  final key = overlayType.trim();
  if (key.isEmpty) return 'Overlay';
  return kOverlayTypeTitles[key] ?? _titleFromSegments(key);
}

/// All built-in overlay types for registry seeding.
Iterable<String> get kAllBuiltinOverlayTypes => kBuiltinOverlayTypes;

File: apps/waddle_controller/src/util/overlayTypeLabel.ts
import type { OverlayTypeSchemaMeta } from '@/storage/configSchemaCache';

const FALLBACK_LABELS: Record<string, string> = {
  shape_rain: 'Shape rain',
  birthday_confetti: 'Birthday confetti',
  bouncing_message: 'Bouncing message',
  falling_images: 'Falling images',
  matrix_rain: 'Matrix rain',
  edge_glow: 'Edge glow',
  floating_balloons: 'Floating balloons',
  static_image: 'Static image',
  digital_clock: 'Digital clock',
  analog_clock: 'Analog clock',
  calendar_month: 'Calendar month',
  calendar_upcoming: 'Calendar upcoming',
  stock_quote: 'Stock quote',
  photo_slideshow: 'Photo slideshow',
};

/** Human-facing label for an overlay type (registry label when available). */
export function overlayTypeLabel(
  overlayType: string | null | undefined,
  meta?: OverlayTypeSchemaMeta | null,
): string {
  const normalized = (overlayType ?? '').trim();
  if (!normalized) return 'unknown';
  const fromLabel = meta?.label?.trim();
  if (fromLabel) return fromLabel;
  return FALLBACK_LABELS[normalized] ?? normalized.replace(/_/g, ' ');
}

export function overlayTypeMetaFor(
  overlayTypes: OverlayTypeSchemaMeta[],
  overlayType: string,
): OverlayTypeSchemaMeta | undefined {
  return overlayTypes.find((m) => m.overlay_type === overlayType);
}
